in_toto/resolver/_resolver.py

```diff
@@ -0,0 +1,61 @@
+"""Resolver interface"""
+
+from abc import ABCMeta, abstractmethod
+from pathspec import PathSpec
+import re
+
+DEFAULT_SCHEME = "default"
+RESOLVER_FOR_URI_SCHEME = {}
+
+
+def _get_scheme(uri):
+  match = re.fullmatch(r"(\w+\:)?(.*)", uri)
+  if not match:
+    raise ValueError(f"Artifact URI '{uri}' could not be parsed")
+  groups = match.groups()
+  if not groups[0]:
+    return DEFAULT_SCHEME
+  return groups[0][:-1]
+
+
+class Resolver(metaclass=ABCMeta):
+  """Interface for resolvers"""
+
+  @classmethod
+  def apply_exclude_patterns(cls, names, exclude_filter):
+    """Exclude matched patterns from passed names."""
+    included = set(names)
+
+    # Assume old way for easier testing
+    if hasattr(exclude_filter, '__iter__'):
+      exclude_filter = PathSpec.from_lines('gitwildmatch', exclude_filter)
+
+    for excluded in exclude_filter.match_files(names):
+      included.discard(excluded)
+
+    return sorted(included)
+
+  @classmethod
+  @abstractmethod
+  def resolve_uri(cls, generic_uri, **kwargs):
+    """Normalize and resolve artifact URIs"""
+    scheme = _get_scheme(generic_uri)
+
+    if scheme not in RESOLVER_FOR_URI_SCHEME:
+      raise ValueError(f"Unsupported in-toto resolver scheme '{scheme}'")
+    resolver = RESOLVER_FOR_URI_SCHEME[scheme]
+
+    return resolver.resolve_uri(generic_uri, **kwargs)
+
+  @classmethod
+  @abstractmethod
+  def hash_artifacts(cls, resolved_uri, hash_algorithms=None, **kwargs):
+    """Return dictionary of hash digests"""
+    scheme = _get_scheme(resolved_uri)
+
+    if scheme not in RESOLVER_FOR_URI_SCHEME:
+      raise ValueError(f"Unsupported in-toto resolver scheme '{scheme}'")
+    resolver = RESOLVER_FOR_URI_SCHEME[scheme]
+
+    return resolver.hash_artifacts(
+        resolved_uri, hash_algorithms=hash_algorithms, **kwargs)
```

Comment:
`resolve_uri` is currently a NOOP in other resolvers since I haven't implemented the function properly. One future application, for example, is having it be able to use a wildcard character (`*`) to get multiple docker containers. It could be squashed into `hash_artifacts` but I thought it might be best to have them be separate.

As for ITE specification, I think we can get away with either one or the other. I went with the latter since that's what we had done earlier, but I could see a case be made for the former since it will ease the burden on the resolvers.

Reply:
I think we should expose both. We have scenarios where `get_hashable_representation` may not be worth the trouble. Consider recording a git or ostree object, for example. We could argue just having their respective hashes suffice. On the other hand, with less definite systems, we probably want to expose both as a means to enable users to examine the data directly when necessary.

Reply:
I implemented a method that would allow this. Developers either implement `get_hashable_representation` or `hash_artifact`.

Reply:
I have consolidated both into a singular api that asks for a `hash_dict`.